File: backend/app/schemas/budget.py
"""Budget schemas for API validation."""
from datetime import datetime as datetime_type
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class BudgetBase(BaseModel):
    """Base budget schema."""
    category_id: int = Field(..., gt=0, description="Category ID")
    year: int = Field(..., ge=2000, le=2100, description="Budget year")
    month: int = Field(..., ge=1, le=12, description="Budget month (1-12)")
    amount: Decimal = Field(..., ge=0, description="Budgeted amount")


class BudgetCreate(BudgetBase):
    """Schema for creating a budget."""
    pass


class BudgetUpdate(BaseModel):
    """Schema for updating a budget (all fields optional for partial updates)."""
    amount: Optional[Decimal] = Field(default=None, ge=0)


class BudgetResponse(BudgetBase):
    """Schema for budget response."""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    created_at: datetime_type
    updated_at: datetime_type


class BudgetWithCategory(BudgetResponse):
    """Schema for budget response with category details."""
    category_name: Optional[str] = None
    category_emoji: Optional[str] = None


class SubcategorySummary(BaseModel):
    """Schema for subcategory monthly summary."""
    subcategory_id: int
    subcategory_name: str
    actual: float
    periods: list[float]

class CategorySummary(BaseModel):
    """Schema for category monthly summary."""
    category_id: int
    category_name: str
    emoji: Optional[str]
    budget: float
    actual: float
    percentage: float
    periods: list[float]
    subcategories: list[SubcategorySummary] = []

class MonthlySummaryResponse(BaseModel):
    """Schema for monthly summary with budget vs actual."""
    year: int
    month: int
    categories: list[CategorySummary]
    total_budget: Decimal
    total_actual: Decimal
    period_boundaries: list[int]  # e.g., [7, 14, 21, 31]

File: SlaveOfCapitalism/SlaveOfCapitalism/Models/Budget.swift
import Foundation

struct BudgetResponse: Codable, Identifiable, Sendable {
    let id: Int
    let categoryId: Int
    let year: Int
    let month: Int
    let amount: Decimal
    let createdAt: String
    let updatedAt: String
}

struct BudgetWithCategory: Codable, Identifiable, Sendable {
    let id: Int
    let categoryId: Int
    let year: Int
    let month: Int
    let amount: Decimal
    let createdAt: String
    let updatedAt: String
    let categoryName: String?
    let categoryEmoji: String?
}

struct BudgetCreate: Codable, Sendable {
    let categoryId: Int
    let year: Int
    let month: Int
    let amount: Decimal
}

struct BudgetUpdate: Codable, Sendable {
    var amount: Decimal?
}

struct CategorySummary: Codable, Identifiable, Sendable {
    let categoryId: Int
    let categoryName: String
    let emoji: String?
    let color: String?
    let budget: Double
    let actual: Double
    let percentage: Double
    let periods: [Double]
    let subcategories: [SubcategorySummary]

    var id: Int { categoryId }
}

struct SubcategorySummary: Codable, Identifiable, Sendable {
    let subcategoryId: Int
    let subcategoryName: String
    let actual: Double
    let periods: [Double]

    var id: Int { subcategoryId }
}

struct MonthlySummaryResponse: Codable, Sendable {
    let year: Int
    let month: Int
    let categories: [CategorySummary]
    let totalBudget: Decimal
    let totalActual: Decimal
    let periodBoundaries: [Int]
}

struct DailyCategoryData: Codable, Identifiable, Sendable {
    let categoryId: Int
    let categoryName: String
    let emoji: String?
    let color: String?
    let budget: Double
    let dailyAmounts: [Double]
    let subcategories: [DailySubcategoryData]

    var id: Int { categoryId }
}

struct DailySubcategoryData: Codable, Identifiable, Sendable {
    let subcategoryId: Int
    let subcategoryName: String
    let dailyAmounts: [Double]

    var id: Int { subcategoryId }
}

struct DailySummaryResponse: Codable, Sendable {
    let year: Int
    let month: Int
    let daysInMonth: Int
    let categories: [DailyCategoryData]
}
